File: gears/hypoid/mesh.py
"""Skew-axis placement and rolling arithmetic for a hypoid pair."""

from __future__ import annotations

import math

from ..placement import (
    Matrix3,
    angle_between,
    angular_velocity_ratio,
    apply,
    gear_mate_ratio,
    matmul,
    rot_y,
    rot_z,
)
from .geometry import HypoidSetGeometry, contact_azimuths

__all__ = [
    "Matrix3", "angle_between", "angular_velocity_ratio", "axis_offset",
    "contact_azimuths", "contact_point", "gear_clocking",
    "gear_contact_point", "gear_mate_ratio", "gear_placement",
    "gear_translation", "pinion_placement", "skew_axis_distance",
]


def pinion_placement(geo: HypoidSetGeometry) -> Matrix3:
    """Clock the pinion's mean tooth space onto the skew contact point."""
    theta1, _ = contact_azimuths(geo)
    return rot_z(theta1)


def gear_clocking(geo: HypoidSetGeometry) -> float:
    """Put a wheel tooth centre at its solved skew contact azimuth."""
    _, target = contact_azimuths(geo)
    tau = 2.0 * math.pi / geo.gear.z
    nearest = round(target / tau - 0.5)
    residual = target - (nearest + 0.5) * tau
    return 0.0 if abs(residual) < 1e-9 else residual % tau


def gear_placement(shaft_angle: float, clocking: float) -> Matrix3:
    return matmul(rot_y(shaft_angle), rot_z(clocking))


def gear_translation(geo: HypoidSetGeometry) -> tuple[float, float, float]:
    """Align both solved mean pitch points while retaining the shaft offset."""
    theta1, theta2 = contact_azimuths(geo)
    a, b = geo.pinion, geo.gear
    pinion_point = (
        a.pitch_radius * math.cos(theta1),
        a.pitch_radius * math.sin(theta1),
        a.cone_distance * math.cos(a.pitch_angle),
    )
    gear_local = (
        b.pitch_radius * math.cos(theta2),
        b.pitch_radius * math.sin(theta2),
        b.cone_distance * math.cos(b.pitch_angle),
    )
    gear_rotated = apply(rot_y(geo.params.sigma), gear_local)
    return tuple(pinion_point[i] - gear_rotated[i] for i in range(3))


def axis_offset(geo: HypoidSetGeometry) -> float:
    return abs(geo.params.offset)


def contact_point(geo: HypoidSetGeometry) -> tuple[float, float, float]:
    """Mean contact point in the pinion frame, useful for pure tests."""
    m = geo.pinion
    theta1, _ = contact_azimuths(geo)
    return (
        m.pitch_radius * math.cos(theta1),
        m.pitch_radius * math.sin(theta1),
        m.cone_distance * math.cos(m.pitch_angle),
    )


def gear_contact_point(geo: HypoidSetGeometry) -> tuple[float, float, float]:
    """The wheel's mean pitch point after skew placement."""
    _, theta2 = contact_azimuths(geo)
    m = geo.gear
    local = (
        m.pitch_radius * math.cos(theta2),
        m.pitch_radius * math.sin(theta2),
        m.cone_distance * math.cos(m.pitch_angle),
    )
    rotated = apply(rot_y(geo.params.sigma), local)
    translation = gear_translation(geo)
    return tuple(rotated[i] + translation[i] for i in range(3))


def skew_axis_distance(origin1, axis1, origin2, axis2) -> float:
    """Shortest distance between two non-parallel 3D axis lines."""
    cross = (
        axis1[1] * axis2[2] - axis1[2] * axis2[1],
        axis1[2] * axis2[0] - axis1[0] * axis2[2],
        axis1[0] * axis2[1] - axis1[1] * axis2[0],
    )
    norm = math.sqrt(sum(value * value for value in cross))
    if norm < 1e-12:
        delta = tuple(origin2[i] - origin1[i] for i in range(3))
        projection = sum(delta[i] * axis1[i] for i in range(3))
        radial = tuple(delta[i] - projection * axis1[i] for i in range(3))
        return math.sqrt(sum(value * value for value in radial))
    delta = tuple(origin2[i] - origin1[i] for i in range(3))
    return abs(sum(delta[i] * cross[i] for i in range(3))) / norm

File: cpp/src/core/hypoid/mesh.cpp
#include "core/hypoid/mesh.hpp"

#include "core/placement/placement.hpp"

#include <cmath>

namespace geargen::core::hypoid::mesh {

Matrix3 pinion_placement(const SetGeometry& geo) noexcept
{
    const auto [theta, unused] = contact_azimuths(geo);
    (void)unused;
    return placement::rot_z(theta);
}

double gear_clocking(const SetGeometry& geo) noexcept
{
    const auto [unused, target] = contact_azimuths(geo);
    (void)unused;
    const double tau = kTau / geo.gear.z;
    const double nearest = std::round(target / tau - 0.5);
    const double residual = target - (nearest + 0.5) * tau;
    if (std::abs(residual) < 1e-9) return 0.0;
    const double wrapped = std::fmod(residual, tau);
    return wrapped < 0.0 ? wrapped + tau : wrapped;
}

Matrix3 gear_placement(double shaft_angle_rad, double clocking_rad) noexcept
{
    return placement::matmul(placement::rot_y(shaft_angle_rad),
                             placement::rot_z(clocking_rad));
}

Point3 contact_point(const SetGeometry& geo)
{
    const auto [theta, unused] = contact_azimuths(geo);
    (void)unused;
    const auto& m = geo.pinion;
    return {m.pitch_radius_mm * std::cos(theta),
            m.pitch_radius_mm * std::sin(theta),
            m.cone_distance_mm * std::cos(m.pitch_angle_rad)};
}

Point3 gear_translation(const SetGeometry& geo)
{
    const auto [theta1, theta2] = contact_azimuths(geo);
    const auto& a = geo.pinion;
    const auto& b = geo.gear;
    const Point3 pinion_point{a.pitch_radius_mm * std::cos(theta1),
                              a.pitch_radius_mm * std::sin(theta1),
                              a.cone_distance_mm * std::cos(a.pitch_angle_rad)};
    const Point3 local{b.pitch_radius_mm * std::cos(theta2),
                       b.pitch_radius_mm * std::sin(theta2),
                       b.cone_distance_mm * std::cos(b.pitch_angle_rad)};
    const Point3 rotated = placement::apply(placement::rot_y(geo.parameters.sigma()), local);
    return {pinion_point.x - rotated.x, pinion_point.y - rotated.y,
            pinion_point.z - rotated.z};
}

Point3 gear_contact_point(const SetGeometry& geo)
{
    const auto [unused, theta2] = contact_azimuths(geo);
    (void)unused;
    const auto& b = geo.gear;
    const Point3 local{b.pitch_radius_mm * std::cos(theta2),
                       b.pitch_radius_mm * std::sin(theta2),
                       b.cone_distance_mm * std::cos(b.pitch_angle_rad)};
    const Point3 rotated = placement::apply(placement::rot_y(geo.parameters.sigma()), local);
    const Point3 translation = gear_translation(geo);
    return rotated + translation;
}

double axis_offset(const SetGeometry& geo) noexcept
{
    return std::abs(geo.parameters.offset);
}

double skew_axis_distance(Point3 origin1, Point3 axis1, Point3 origin2,
                          Point3 axis2) noexcept
{
    const Point3 cross{axis1.y * axis2.z - axis1.z * axis2.y,
                       axis1.z * axis2.x - axis1.x * axis2.z,
                       axis1.x * axis2.y - axis1.y * axis2.x};
    const double magnitude = norm(cross);
    const Point3 delta = origin2 - origin1;
    if (magnitude < 1e-12) {
        const double projection = dot(delta, axis1);
        return norm(delta - axis1 * projection);
    }
    return std::abs(dot(delta, cross)) / magnitude;
}

} // namespace geargen::core::hypoid::mesh
